pkg/cvo/status.go

```diff
@@ -2,130 +2,222 @@ package cvo
+	config.Status.Current = optr.currentVersion()
+
+	_, _, err := resourceapply.ApplyClusterVersionStatusFromCache(optr.cvoConfigLister, optr.client.ConfigV1(), config)
+	return err
+}
+
+func (optr *Operator) syncAvailableStatus(config *cvv1.ClusterVersion) error {
+	now := metav1.Now()
+	version := optr.currentVersionString(config)
+	resourcemerge.SetOperatorStatusCondition(&config.Status.Conditions, osv1.ClusterOperatorStatusCondition{
+		Type:    osv1.OperatorAvailable,
+		Status:  osv1.ConditionTrue,
+		Message: fmt.Sprintf("Done applying %s", version),
+
+		LastTransitionTime: now,
+	})
+	resourcemerge.SetOperatorStatusCondition(&config.Status.Conditions, osv1.ClusterOperatorStatusCondition{
+		Type:    osv1.OperatorProgressing,
+		Status:  osv1.ConditionFalse,
+		Message: fmt.Sprintf("Cluster version is %s", version),
+
+		LastTransitionTime: now,
 	})
-	resourcemerge.SetOperatorStatusCondition(&status.Status.Conditions, osv1.ClusterOperatorStatusCondition{Type: osv1.OperatorProgressing, Status: osv1.ConditionFalse})
-	resourcemerge.SetOperatorStatusCondition(&status.Status.Conditions, osv1.ClusterOperatorStatusCondition{Type: osv1.OperatorFailing, Status: osv1.ConditionFalse})
+	resourcemerge.SetOperatorStatusCondition(&config.Status.Conditions, osv1.ClusterOperatorStatusCondition{Type: osv1.OperatorFailing, Status: osv1.ConditionFalse, LastTransitionTime: now})
+
+	config.Status.Generation = config.Generation
 
-	_, _, err := resourceapply.ApplyOperatorStatusFromCache(optr.clusterOperatorLister, optr.client.OperatorstatusV1(), status)
+	_, _, err := resourceapply.ApplyClusterVersionStatusFromCache(optr.cvoConfigLister, optr.client.ConfigV1(), config)
 	return err
 }
 
-// syncDegradedStatus updates the OperatorStatus to Degraded.
+// syncDegradedStatus updates the ClusterVersion status to Degraded.
 // if ierr is nil, return nil
 // if ierr is not nil, update OperatorStatus as Degraded and return ierr
-func (optr *Operator) syncDegradedStatus(ierr error) error {
+func (optr *Operator) syncDegradedStatus(config *cvv1.ClusterVersion, ierr error) error {
 	if ierr == nil {
 		return nil
 	}
 
-	status := &osv1.ClusterOperator{
+	clusterVersion := &cvv1.ClusterVersion{
 		ObjectMeta: metav1.ObjectMeta{
-			Namespace: optr.namespace,
-			Name:      optr.name,
-		},
-		Status: osv1.ClusterOperatorStatus{
-			Version:   version.Raw,
-			Extension: runtime.RawExtension{},
+			Name: optr.name,
 		},
 	}
-	resourcemerge.SetOperatorStatusCondition(&status.Status.Conditions, osv1.ClusterOperatorStatusCondition{Type: osv1.OperatorAvailable, Status: osv1.ConditionFalse})
-	resourcemerge.SetOperatorStatusCondition(&status.Status.Conditions, osv1.ClusterOperatorStatusCondition{Type: osv1.OperatorProgressing, Status: osv1.ConditionFalse})
-	resourcemerge.SetOperatorStatusCondition(&status.Status.Conditions, osv1.ClusterOperatorStatusCondition{
-		Type: osv1.OperatorFailing, Status: osv1.ConditionTrue,
+	// try to reuse the most recent status if available
```

Comment:
why is degraded different compared to available, progresing.

Reply:
Right now called from two different spots - inside the sync loop, and outside when a generic error happens.

I’m considering a follow up where I add better error output for sync and release access.  I can refactor this there (we might want a structured error set for displaying more succinct errors)

Reply:
#47 has most of these fixes, and clarifies naming.